src/main/kotlin/it/pagopa/wallet/services/MigrationService.kt

```diff
@@ -0,0 +1,90 @@
+import it.pagopa.wallet.config.WalletMigrationConfig
+import it.pagopa.wallet.domain.migration.WalletPaymentManager
+import it.pagopa.wallet.domain.migration.WalletPaymentManagerRepository
+import it.pagopa.wallet.domain.wallets.*
+import it.pagopa.wallet.repositories.LoggingEventRepository
+import it.pagopa.wallet.repositories.WalletRepository
+import it.pagopa.wallet.util.UniqueIdUtils
+import java.time.Instant
+import java.util.*
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.dao.DuplicateKeyException
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmptyDeferred
+import reactor.kotlin.core.publisher.toMono
+
+@Service
+class MigrationService(
+    private val walletPaymentManagerRepository: WalletPaymentManagerRepository,
+    private val walletRepository: WalletRepository,
+    private val loggingEventRepository: LoggingEventRepository,
+    private val uniqueIdUtils: UniqueIdUtils,
+    walletMigrationConfig: WalletMigrationConfig
+) {
+
+    private val cardPaymentMethodId =
+        PaymentMethodId(UUID.fromString(walletMigrationConfig.cardPaymentMethodId))
+    private val logger: Logger = LoggerFactory.getLogger(this.javaClass)
+
+    fun initializeWalletByPaymentManager(
+        paymentManagerWalletId: String,
+        userId: UserId
+    ): Mono<Wallet> {
+        logger.info(
+            "Initialize wallet for PaymentManager Id: $paymentManagerWalletId and userId: $userId"
+        )
+        val now = Instant.now()
+        return walletPaymentManagerRepository
+            .findByWalletPmId(paymentManagerWalletId)
+            .switchIfEmptyDeferred { createMigrationData(paymentManagerWalletId) }
+            .flatMap { createWalletByPaymentManager(it, userId, cardPaymentMethodId, now) }
+            .doOnError { logger.error(it.message) }
+            .toMono()
+    }
+
+    private fun createWalletByPaymentManager(
+        migration: WalletPaymentManager,
+        userId: UserId,
+        paymentMethodId: PaymentMethodId,
+        creationTime: Instant
+    ): Mono<Wallet> {
+        val newWallet =
+            Wallet(
+                id = migration.walletId,
+                userId = userId,
+                contractId = migration.contractId,
+                status = WalletStatusDto.CREATED,
+                paymentMethodId = paymentMethodId,
+                creationDate = creationTime,
+                updateDate = creationTime,
+                version = 0,
+            )
+        return walletRepository
+            .save(newWallet.toDocument())
+            .map { LoggedAction(it.toDomain(), WalletAddedEvent(it.id)) }
+            .flatMap { it.saveEvents(loggingEventRepository) }
+            .onErrorResume(DuplicateKeyException::class.java) {
+                walletRepository.findById(migration.walletId.value.toString()).map { it.toDomain() }
+            }
```

Comment:
```suggestion
        return walletRepository
            .findById(migration.walletId.value.toString())
            .switchIfEmpty(
                walletRepository
                    .save(newWallet.toDocument())
                    .map { LoggedAction(it.toDomain(), WalletAddedEvent(it.id)) }
                    .flatMap { it.saveEvents(loggingEventRepository) }
                    .map { it.toDocument() }
            )
            .map { it.toDomain() }
```
Generally speaking find by id is a less RU consumption operation than a save (that can go wrong because of duplicate key).
Maybe this piece of code can be refactored to perform a findById and then perform wallet save operation iff no wallet is found for the id?
Something like this, what do you think?

Reply:
I'm agree about RU consumption. This api has to create a new wallet from PM, in 90% of the cases I expect the wallet has to be created (unless retry). So by adding a find the operations to create a wallet would be 2, a find and a save. In case of retry on an already created wallet on the other hand I agree that the cost is higher since a save and then a find is done here. I expect that the wallets in this scenario are all to be created since they migrate from PM. What do you think?

Reply:
Yes, I was missing this context, perfect, we can leave as is since read operation should practically only happen in retry case 🚀